src/xpcc/architecture/platform/driver/adc/stm32f0/adc.hpp.in

```diff
@@ -0,0 +1,260 @@
+// coding: utf-8
+/* Copyright (c) 2013, Roboterclub Aachen e.V.
+* All Rights Reserved.
+*
+* The file is part of the xpcc library and is released under the 3-clause BSD
+* license. See the file `LICENSE` for the full license governing this code.
+*/
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC_STM32F0_ADC_HPP
+#define XPCC_STM32F0_ADC_HPP
+
+#include <stdint.h>
+#include "../../../type_ids.hpp"
+#include "../../../device.hpp"
+#include <xpcc/architecture/interface/register.hpp>
+
+namespace xpcc
+{
+namespace stm32
+{
+/**
+ * Analog/Digital-Converter module (ADC1).
+ *
+ * The 12-bit ADC is a successive approximation analog-to-digital
+ * converter. It has up to 19 multiplexed channels allowing it measure
+ * signals from 16 external and three internal sources.
+ * The result of the ADC is stored in a left-aligned or right-aligned
+ * 16-bit data register.
+ *
+ * This API is designed for the internal ADCs of STM32F0x1/STM32F0x2/STM32F0x8
```

Comment:
I only looked briefly into the ADC code of STM32F3. Isn't there a chance to integrate it to the F3 code? There seems to be lots of duplicated code. I know that the ADC interface in xpcc is not the cleanest and this could be a "works for me"-solution.

Reply:
Are the subfolders in the ADC some kind of more-specific alternative? So if there's a stm32fx folder pick it, otherwise fallback to stm32? If so, what was the reason to do it this way in the first place? Should everything be merged into a single master template in the stm32 folder? Merging templates will fill it with %ifs and make it hard to read later, I think. Is hierarchy a solution? Put common code in a base class, defining an API and then subclassing and overloading the methods that will be specific for the device. The later seems like a better solution to me, but I dont have that much experience with µC to know how it affects performance.

Reply:
Lets leave this as it is right now. The xpcc code generation system is very fragile and didn't scale gracefully to so many implementations.
We've got much better tools in modm for fine-grain control over code generation. My comparison tools for the STM32 also suggest that there is still significant difference between F3 and F0 ADC implementations.

Reply:
To my understanding this grew historically. First, the STM32F4 was supported (code in `stm32`. There is no stm32f4 directory). Then, it was discovered that F3 is quite different and new code was created in a new subdirectory. L4 was then later add to the F3, because they share much. F7 and F2 are similar to F4, so it was then added to this driver.

There is no fallback in place. The ADC driver is explicitly selected in the [device XML](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/devices/stm32/stm32l431_433_443-c_k_r_v-b_c.xml#L103).

All drivers should not be merged into one file. This of course will make a `%if` mess. ADCs which are somehow similar should be put into one file (F2, F4, F7) / (F3, L4). To which group would F0 fit? From which template did you take your driver?

Reply:
I took the files from stm32f3, but just to have a basic template and class definition I could follow. Must of the code was rewritten taking, they are not really compatible.